File: src/leetCodeProblems/StacksAndQueues/QueueRawImpl.java
package leetCodeProblems.StacksAndQueues;

/**
 * Reference - https://www.educba.com/stack-vs-queue/
 * @author anshul.agrawal
 *
 */

public class QueueRawImpl {
	private int maxSize;
	private String[] stackArray;
	private int first = 0;
	private int last = -1;

	public QueueRawImpl(int s) {
		maxSize = s;
		stackArray = new String[maxSize];
	}

	public void push(String j) {
		last += 1;
		stackArray[last] = j;
	}

	public String pop() {
		String temp = stackArray[first];
		first += 1;
		return temp;
	}

	public String peek() {
		return stackArray[first];
	}

	public boolean isEmpty() {
		return (last == -1);
	}

	public boolean isFull() {
		return (last == maxSize - 1);
	}

}

File: src/leetCodeProblems/StacksAndQueues/QueueRawImpl.go
package main

//import "log"

type Queue []int

func (s *Queue) IsEmpty() bool {
	return len(*s) == 0
}
func (s *Queue) Push(num int) {
	*s = append(*s, num)
}
func (s *Queue) Len() int {
	return len(*s)
}
func (s *Queue) Pop() (int, bool) {
	if s.IsEmpty() {
		return -1, false
	} else {
		elem := (*s)[0]
		*s = (*s)[1:]
		return elem, true
	}
}

// func main() {
// 	queue := new(Queue)
// 	queue.Push(1)
// 	queue.Push(2)
// 	queue.Push(3)
// 	log.Println(queue.Pop())
// 	log.Println(queue.Pop())
// }
